File: 2024/d2/main.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

bool is_safe(int *line, size_t count, int skip_index) {
    bool desc = false, desc_set = false;

    int start = skip_index == 0 ? 2 : 1;
    int first = line[start - 1];
    int prev = first;

    for (int i = start; i < count; i++) {
        if (i == skip_index) continue;
        int val = line[i];

        if (!desc_set) {
            desc = val < first;
            desc_set = true;
        }
        if ((desc && val > prev)|| (!desc && val < prev)) {
            return false;
        }
        int diff = desc ? prev - val : val - prev;
        if (!(1 <= diff && diff <= 3)) return false;
    
        prev = val;
    }

    return true;
}

size_t parse_nums(char *line, int *buff) {
    char *curr = line;
    size_t count = 0;

    do {
        int value = (int)strtol(curr, &curr, 10);
        buff[count++] = value;
    } while (*curr != '\0');

    return count;
    
}

int main() {
FILE *file = fopen("input.txt", "r");

    fseek(file, 0, SEEK_END);
    long size = ftell(file);
    fseek(file, 0, SEEK_SET);
    char contents[size];
    fread(contents, 1, size, file);
    fclose(file);

    char *state = contents;
    char *line;
    int safe_count = 0;
    while ((line = strtok_r(state, "\n", &state)) != NULL) {
        bool safe = false;
        int buff[128];
        size_t count = parse_nums(line, buff);

        if (is_safe(buff, count, -1)) safe = true;
        else {
            for (int i = 0; i < count; i++) {
                if (is_safe(buff, count, i)){
                    safe = true;
                    break;
            }
            }
        }

        if (safe) {
            safe_count++;
            printf("safe\n");
        } else printf("unsafe\n");
    }

    printf("result: %d\n", safe_count);
    return 0;
}

File: 2024/d2/main.zig
const std = @import("std");
const mem = std.mem;
const fmt = std.fmt;
const print = std.debug.print;

fn isSafe(line: []i32, skipIndex: i32) bool {
    var desc = false;
    var descIsSet = false;

    const start: u8 = if (skipIndex == 0) 2 else 1;
    const first = line[start - 1];
    var prev = first;

    var i: u32 = start;
    for (line[start..]) |val| {
        defer i += 1;
        if (i == skipIndex) continue;
        if (!descIsSet) {
            desc = val < first;
            descIsSet = true;
        }

        if (desc and val > prev) {
            return false;
        } else if (!desc and val < prev) {
            return false;
        }
        const diff = @abs(val - prev);
        if (!(1 <= diff and diff <= 3)) {
            return false;
        }
        prev = val;
    }
    return true;
}

fn parseNums(line: []u8, alloc: mem.Allocator, comptime maxSize: i32) ![]i32 {
    var buffer: [maxSize]i32 = undefined;
    var tokens = mem.splitSequence(u8, line, " ");
    var count: u32 = 0;
    while (tokens.next()) |tok| : (count += 1) {
        buffer[count] = try fmt.parseInt(i32, tok, 10);
    }

    const nums: []i32 = try alloc.alloc(i32, count);
    mem.copyForwards(i32, nums, buffer[0..count]);
    return nums;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){}; // gpa is of type GeneralPurposeAllocator()
    const alloc = gpa.allocator();

    const file = try std.fs.cwd().openFile("input.txt", .{});
    defer file.close();

    const contents = try file.readToEndAlloc(alloc, 100_000);
    defer alloc.free(contents);

    var lines = mem.splitSequence(u8, contents, "\n");
    var safeCount: i32 = 0;
    while (lines.next()) |line| {
        if (std.mem.eql(u8, line, "")) continue;

        var safe = false;
        const slice = try parseNums(@constCast(line), alloc, 64);
        defer alloc.free(slice);
        // check if safe
        if (isSafe(slice, -1)) {
            safe = true;
        } else {
            // brute force remove numbers
            for (0..slice.len) |i| {
                if (isSafe(slice, @intCast(i))) {
                    safe = true;
                    break;
                }
            }
        }
        if (safe) {
            safeCount += 1;
            print("safe: {d}\n", .{slice});
        } else {
            print("unsafe: {d}\n", .{slice});
        }
    }

    print("{}\n", .{safeCount});
}
